CAGRA pad dataset for 128bit vectorized load

tfeher opened the pull request:
This PR adds padding to the dataset (if necessary) to make reading any of its rows compatible with 128bit vectorized loads. This change also enables handling arbitrary number of input features (before this PR each row had to be at least 64bit aligned, which constrained the acceptable number of input features).

Fixes #1458.

With this change, it is sufficient to keep a single "load type" specialization for the search kernels, which shall cut the binary size by half (#1459).

tfeher commented:
Marked as breaking change because it removes the `load_bit_length` search parameter.

tfeher commented:
There is still a bug with the new tests, otherwise it is ready for review.

enp1s0 commented:
The solution of the compilation error is:

https://github.com/tfeher/raft/blob/fb175783a6cb6b9d0e56f0c18bd84c269fce9bb1/cpp/include/raft/neighbors/detail/cagra/search_multi_cta.cuh#L231
```diff
231:      dataset_size,
+         dataset_ld,
232:      result_buffer_size,
```

enp1s0 commented:
It would be better to support a padded dataset in the argument of `cagra::sort_knn_graph`. What do you think?

enp1s0 reviewed (commented):
The changes look good to me.

One comment:
It would be better to support a padded dataset in the argument of `cagra::sort_knn_graph`. What do you think?

cjnolet commented:
@enp1s0 please fix the conflicts and we can merge this in for release 23.06.

tfeher commented:
I have fixed a bug in the serialization routine, and added more tests. @enp1s0 could you have a quick look again at the changes?

> It would be better to support a padded dataset in the argument of cagra::sort_knn_graph. What do you think?

I think it is a good suggestion, and it might be also useful to allow the constructor of index to accept padded dataset (i.e. strided mdspan). I would prefer to make these changes in a follow up PR.

enp1s0 commented:
Thank you, @tfeher, for fixing it and adding tests. The code looks good to me.

enp1s0 reviewed (approved):
Thank you, @tfeher, for fixing it and adding tests. The code looks good to me.

cjnolet commented on cpp/include/raft/neighbors/detail/cagra/cagra_serialize.cuh:
```diff
@@ -59,7 +60,19 @@ void serialize(raft::resources const& res, std::ostream& os, const index<T, IdxT
+                                  cudaMemcpyDefault,
+                                  resource::get_cuda_stream(res)));
+  resource::sync_stream(res);
+  serialize_mdspan(res, os, host_dataset.view());
```
Mostly a side question, but are we still planning to remove the dataset from the serialization in a future change?

cjnolet commented on cpp/include/raft/neighbors/cagra_types.hpp:
```diff
@@ -179,15 +183,36 @@ struct index : ann::index {
+                 resource::get_cuda_stream(res));
+    } else {
+      // copy with padding
+      RAFT_CUDA_TRY(cudaMemsetAsync(
```
If this is going to be a more common practice, I wonder if we should consider centralizing this somewhere eventually. Probably doesn't need to be done yet, or even in this PR, though.

cjnolet commented on cpp/include/raft/neighbors/detail/cagra/search_multi_cta.cuh:
```diff
@@ -569,7 +546,7 @@ struct search : public search_plan_impl<DATA_T, INDEX_T, DISTANCE_T> {
 
   void operator()(raft::resources const& res,
-                  raft::device_matrix_view<const DATA_T, INDEX_T, row_major> dataset,
+                  raft::device_matrix_view<const DATA_T, INDEX_T, layout_stride> dataset,
```
Would there be any benefit to using a  [padded layout](https://github.com/rapidsai/raft/blob/567dfd74a5c092cec61de6b52ea239ea91b2ed46/cpp/include/raft/core/mdspan.hpp#L58) here or having an overload for it in the public API just to simplify the conversion?

cjnolet reviewed (approved):
LGTM. Just a couple questions and a suggestion for a future improvement overall. Since CAGRA is still experimental, we have some leeway for API changes.

cjnolet commented:
/merge
